cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -128,18 +132,45 @@
   }
 
   @Override
+  protected RuntimeInfo createRuntimeInfo(ProgramController controller, Program program) {
+    RunId twillRunId = ((AbstractTwillProgramController) controller).getTwillRunId();
+    return new SimpleRuntimeInfo(controller, program, twillRunId);
+  }
+
+  @Nullable
+  private synchronized RunId getTwillRunId(RunId runId) {
+    RuntimeInfo runtimeInfo = super.lookup(runId);
+    if (runtimeInfo != null) {
+      return runtimeInfo.getTwillRunId();
+    }
+    return null;
+  }
+
+  private synchronized boolean twillRunIdCached(RunId twillRunId) {
+    for (RuntimeInfo runtimeInfo : runtimeInfos.values()) {
+      if (runtimeInfo.getTwillRunId() != null && runtimeInfo.getTwillRunId().equals(twillRunId)) {
```

Comment:
Assuming the `twillRunId` in the parameter never null, which is ok in this case since this is a private method.